release-23.1: server/license: Make enterprise license check a no-op

spilchen opened the pull request:
Backport 1/1 commits from #131943.

/cc @cockroachdb/release

---

With the deprecation of the core license, there’s no longer a need to check whether enterprise features are enabled—all features are now always enabled. If a license policy violation occurs, such as an expired license, we will throttle connections, but all enterprise features will remain functional. I’ve left the API for checking if enterprise features are enabled in the code, but these functions are now no-ops and can be removed in the future.

I also had to adjust some tests that were explicitly disabling enterprise features, as that’s no longer possible. Some tests were rewritten to fit the new model, while others were removed entirely because they were no longer relevant.

Note: A few tests in `pkg/sql` had to be updated due to the changes in enterprise checks. With all enterprise features enabled, the default isolation level has also changed. Any test that explicitly disabled leases (using TestingDisableTableLeases) needed to apply this setting _after_ starting up the test server. Applying it earlier, as was originally done, caused the test server to fail during startup.

This change will be backported to versions 24.2, 24.1, 23.2, and 23.1.

Epic: CRDB-39988
Closes: CRDB-41998
Release note: none
Release justification: This work is part of the CockroachDB core deprecation.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/132427)
<!-- Reviewable:end -->

rail commented:
git commit 9f02d0e966faa336a2ac10e98359e319b170b950

us-east1-docker.pkg.dev/crl-ci-images/cockroach/acceptance-gss-psql:20241011-130122
us-east1-docker.pkg.dev/crl-ci-images/cockroach/acceptance-gss-python:20241011-130452

spilchen committed:
server/license: Make enterprise license check a no-op

With the deprecation of the core license, there’s no longer a need to
check whether enterprise features are enabled—all features are now
always enabled. If a license policy violation occurs, such as an expired
license, we will throttle connections, but all enterprise features will
remain functional. I’ve left the API for checking if enterprise features
are enabled in the code, but these functions are now no-ops and can be
removed in the future.

I also had to adjust some tests that were explicitly disabling
enterprise features, as that’s no longer possible. Some tests were
rewritten to fit the new model, while others were removed entirely
because they were no longer relevant.

Note: A few tests in pkg/sql had to be updated due to the changes in
enterprise checks. With all enterprise features enabled, the default
isolation level has also changed. Any test that explicitly disabled
leases (using TestingDisableTableLeases) needed to apply this setting
after starting up the test server. Applying it earlier, as was
originally done, caused the test server to fail during startup.

This change will be backported to versions 24.2, 24.1, 23.2, and 23.1.

Epic: CRDB-39988
Closes: CRDB-41998
Release note: none

spilchen commented:
This backport is slightly different than other releases. It required preserving some of the enterprise checks, as around 70 unit tests were failing. The tests relied on the test cluster being set up in a specific way, which was itself controlled by the enterprise check (see testserver.go).
